### ISSUE TITLE
How to create an "actual" fault tolerance system ? 

### ISSUE THREAD

**Issue by dxj910804** [nats-io/nats-streaming-server#712]
I'm looking to create a "reliable" msg system with fault tolerance. 
My understanding of clustered mode and fault tolerance mode is: 
clustered mode with 2 different binaries nats streaming and nats server, running with its own storage.
fault tolerance mode with single binary nats streaming running with embeded nats server, running with shared storage. 
My questions are :  
1,  If I want to run nats streaming server on multiple machines, does this exclude "fault tolerance" mode ? Cause it needs a shared store, or I need to provide a remote shared file/data base. 
2, Is it not possible to have embeded nats servers in this case ? There's an example doing this with multiple external nats servers and nats streaming servers. https://nats.io/documentation/tutorials/nats-streaming-swarm/ 
3, Even with multiple external nats servers and nats streaming servers, if the leader goes done, is it possible to avoid clients to reconnect ? 
4, How to choose between these two modes and why ?

**Comment by kozlovic** [comment]
> clustered mode with 2 different binaries nats streaming and nats server, running with its own storage. fault tolerance mode with single binary nats streaming running with embeded nats server, running with shared storage.

This is not correct. The fact that NATS Streaming can run with an embedded NATS Server or connect to a remote one has nothing to do with  standalone/clustering/FT mode.

> 1, If I want to run nats streaming server on multiple machines, does this exclude "fault tolerance" mode? ? Cause it needs a shared store, or I need to provide a remote shared file/data base.

It is expected to actually run on different machines to provide real fault tolerance ;-). However, yes, it means that a shared storage need to be available.

> 2, Is it not possible to have embeded nats servers in this case ? There's an example doing this with multiple external nats servers and nats streaming servers. https://nats.io/documentation/tutorials/nats-streaming-swarm/

Again, embedded NATS Server or not is not a function of the mode the server is running in. This is just a convenience to not have to start an independent NATS Server, but in some cases you may want to connect to an existing NATS cluster, in which case you would configure the streaming servers to "point" to this cluster.

> 3, Even with multiple external nats servers and nats streaming servers, if the leader goes done, is it possible to avoid clients to reconnect ?

In NATS Streaming, both FT and cluster, there is a leader that handles all traffic. So when leader of FT active goes down, another streaming server will take over. That being said, in term of actual low level NATS connection, a client may or may not be connected to the leader. I would recommend that you look at https://github.com/nats-io/nats-streaming-server#relation-to-nats to have a better understanding of the relation between NATS Streaming and core NATS.

> 4, How to choose between these two modes and why ?

I would say that you would select clustering if you cannot have a shared storage. If you do have access to shared storage, FT is more lightweight and performs better.

**Comment by dxj910804** [comment]
Thanks for prompt reply. 
I agree that "The fact that NATS Streaming can run with an embedded NATS Server or connect to a remote one has nothing to do with standalone/clustering/FT mode." Another fact is that, if we run embeded nats server with nats streaming clustered, then the one running with embeded nats server is the actual single failure node. Is it possible to cluster multiple embeded nats servers and the nats streaming servers ?

**Comment by kozlovic** [comment]
Please check [this](https://github.com/nats-io/nats-streaming-server#clustering-configuration). There is an example of running 3 servers. Look for the paragraph starting with:
```
On host1, this configuration indicates that the server will accept client connections on port 4222. It will accept route connections on port 6222. It creates 2 routes, to host2 and host3 cluster port.
```
The following section shows 3 configs, each one defining the NATS cluster section with port and routes to other nodes.

In above mode, each NATS Streaming server embeds a NATS Server and clusters to each of the other nodes. Clients are NATS connected to any of the servers in the cluster. For clients connected to the server that would fail, they would reconnect to any other server in the cluster.
